Splices with non-`String` type break if there is no splice.

When using #743 with a splice handler that does not return a `String`, like e.g.:
```
def mylit{ body: => Unit / { literal } }: Int = <>
```
we will get a type error when using it *without actually splicing anything*:
```
def main() = {
  println(show(mylit"foo" + 2)) // Cannot typecheck call.
}
```

## Runtime behaviour
Furthermore, if we do a transformation on the `String`, e.g.:
```
def mylit{ body: => Unit / { literal } }: String =
  try { body(); "1" } with literal { l => "0" }
```
it will silently drop the `mylit` for `Strings` without splices:
```
def main() = {
  println(mylit"foo") // prints "foo", not "0"
}
```

After fixing this, please don't forget to add a test to `examples/pos/` 
I'd also like to see the following examples @timsueberkrueb and @marzipankaiser came up with added to the test in order to showcase string interpolation more :)
```scala
effect log(msg: String): Unit

def error { body: => Unit / literal }: Unit / log =
  try body() with literal { x => do log("[ERROR] " ++ x); resume(()) }
def warn { body: => Unit / literal }: Unit / log =
  try body() with literal { x => do log("[WARNING] " ++ x); resume(()) }

def doc { body: => Unit / {literal} }: Unit =
  try body() with literal { _ => resume(()) }

def main() = try {
  doc"This is the doc string for my main function!"

  println(42)
  warn"Frobnicators have been jabberwocked!"

  println(-1)
  error"Stuff went wrong!"
  error"Aborting!"
} with log { msg => println(msg); resume(()) }
```

--- a/effekt/shared/src/main/scala/effekt/RecursiveDescent.scala
+++ b/effekt/shared/src/main/scala/effekt/RecursiveDescent.scala
@@ -1030,10 +1030,9 @@ class RecursiveDescent(positions: Positions, tokens: Seq[Token], source: Source)
   def templateString(): Term =
     nonterminal:
       backtrack(idRef()) ~ template() match {
-        // We do not need to apply any transformation if there are no splices
-        case _ ~ Template(str :: Nil, Nil) => StringLit(str)
-        case _ ~ Template(strs, Nil) => fail("Cannot occur")
-        // s"a${x}b${y}" ~> s { do literal("a"); do splice(x); do literal("b"); do splice(y) }
+        // We do not need to apply any transformation if there are no splices _and_ no custom handler id is given
+        case None ~ Template(str :: Nil, Nil) => StringLit(str)
+        // s"a${x}b${y}" ~> s { do literal("a"); do splice(x); do literal("b"); do splice(y); return () }
         case id ~ Template(strs, args) =>
           val target = id.getOrElse(IdRef(Nil, "s"))
           val doLits = strs.map { s =>

--- a/examples/pos/string_interpolation.check
+++ b/examples/pos/string_interpolation.check
@@ -1,2 +1,3 @@
 GET https://api.effekt-lang.org/users/effekt/resource/42
 Fix point combinator: \ f -> (\ x -> f x x) \ x -> f x x
+12

--- a/examples/pos/string_interpolation.effekt
+++ b/examples/pos/string_interpolation.effekt
@@ -28,6 +28,15 @@ def pretty { prog: () => Unit / {literal, splice[Expr]} }: String = {
   }
 }
 
+def len { prog: () => Unit / {literal} }: Int = {
+  try {
+    prog()
+    0
+  } with literal { s =>
+    s.length
+  }
+}
+
 def main() = {
   val domain = "https://api.effekt-lang.org"
   val user = "effekt"
@@ -36,4 +45,6 @@ def main() = {
 
   val fixpoint = Abs("f", App(Abs("x", App(Var("f"), App(Var("x"), Var("x")))), Abs("x", App(Var("f"), App(Var("x"), Var("x"))))))
   println(pretty"Fix point combinator: ${fixpoint}")
+
+  println(show(len"hello, world"))
 }

--- a/examples/pos/string_interpolation_literal.check
+++ b/examples/pos/string_interpolation_literal.check
@@ -1,0 +1,5 @@
+42
+[WARNING] Frobnicators have been jabberwocked!
+-1
+[ERROR] Stuff went wrong!
+[ERROR] Aborting!

--- a/examples/pos/string_interpolation_literal.effekt
+++ b/examples/pos/string_interpolation_literal.effekt
@@ -1,0 +1,21 @@
+effect log(msg: String): Unit
+
+def error { body: => Unit / literal }: Unit / log =
+  try body() with literal { x => do log("[ERROR] " ++ x); resume(()) }
+
+def warn { body: => Unit / literal }: Unit / log =
+  try body() with literal { x => do log("[WARNING] " ++ x); resume(()) }
+
+def doc { body: => Unit / {literal} }: Unit =
+  try body() with literal { _ => resume(()) }
+
+def main() = try {
+  doc"This is the doc string for my main function!"
+
+  println(42)
+  warn"Frobnicators have been jabberwocked!"
+
+  println(-1)
+  error"Stuff went wrong!"
+  error"Aborting!"
+} with log { msg => println(msg); resume(()) }

